### ISSUE TITLE
Add ffmpeg

### ISSUE THREAD

**Issue by prabirshrestha** [holzschu/a-shell#8]
Would like to use this in conjunction with Siri Shortcuts to convert my gopro videos to smaller size on iPad given that iPad itself is pretty powerful.

**Comment by suliveevil** [comment]
+1

**Comment by dbellingroth** [comment]
+1

**Comment by erikhermesman** [comment]
+1 Would be great for added functionality in youtube-dl

**Comment by holzschu** [comment]
Hi, 
ffmpeg is now available as a binary to download. From here: https://github.com/holzschu/a-Shell-commands , download `ffmpeg.wasm`, place it in `~/Documents/bin` and type `ffmpeg` on the command line. 

It takes a lot of memory, so it might crash on smaller iPhones. Also it can take a long time to process a large video.

**Comment by suliveevil** [comment]
Great Package！

**Comment by erikhermesman** [comment]
Thank you!

**Comment by ifuchs** [comment]
ffmpeg crashes my iPad Pro (9.7-inch) and my iPhone 7.

**Comment by holzschu** [comment]
Yes, ffmpeg is a bit memory intensive (the binary itself is 87 MB), so just loading it in memory requires a powerful iPad, with few other memory-intensive apps running. After that, it can take a long time to complete the task, too.

**Comment by erikhermesman** [comment]
When I ‘ls’ into ~/Documents/bin I get a no such document/directory error and when I try to mv the ffmpeg.wasm file into that directory I also get an error. Any ideas what I’m doing wrong? Thanks.

**Comment by holzschu** [comment]
You might need to create the directory first if it doesn't exist, with `mkdir`.

**Comment by suliveevil** [comment]
```shell
mkdir -p ~/Documents/bin
mv ffmpeg.wasm ~/Documents/bin/
```

**Comment by erikhermesman** [comment]
Thanks! Got it working (didn’t need the “cd ~”). Unfortunately though, like one of the above posts, ffmpeg crashes when running it on my iPad and iPhone 7. Haven’t tried it with youtube-dl yet, which is mainly what I want it for.

**Comment by atnbueno** [comment]
What am I doing wrong?
```
$ mkdir ~/Documents/bin
$ curl https://github.com/holzschu/a-Shell-commands/releases/download/0.1/ffmpeg.wasm -o ~/Documents/bin/ffmpeg.wasm
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   624  100   624    0     0   1318      0 --:--:-- --:--:-- --:--:--  1319
$ ffmpeg
wasm: Error at line 59, column 37: Error: Unreachable code should not be executed (evaluating 'wasm.lowerI64Imports(retptr, passArray8ToWasm(wasm_binary), WASM_VECTOR_LEN)')
$ ls -la ~/Documents/bin
total 8
drwxr-xr-x  3 mobile  mobile   96 Jul 28  2020 .
drwxr-xr-x  3 mobile  mobile   96 Jul 28  2020 ..
-rw-r--r--  1 mobile  mobile  624 Jul 28  2020 ffmpeg.wasm
$
```

**Comment by holzschu** [comment]
I don't know what went wrong, but ffmpeg.wasm is 88 MB. If the file that was dowloaded is only 624 bytes, then something went wrong during the download. `cat ~/Documents/bin/ffmpeg.wasm` will probably tell you what.

**Comment by lukefor** [comment]
Given that ffmpeg seems to be a popular use case for A-Shell, and is performance critical, is it possible to compile it natively as part of A-Shell itself, rather than a wasm binary?

**Comment by faraixyz** [comment]
Is it possible for ffmpeg.wasm to include encoders for h264 and h265? So far it's helpful for changing the file format, but it would be nice to re-encode videos to make them smaller.

**Comment by holzschu** [comment]
Hi, the latest release of a-Shell has native ffmpeg, with hardware-accelerated h264 encoding. You will have to remove the webAssembly version in order to access the native version.

**Comment by shyam1312** [comment]
I installed latest version a-Shell on my iPad . How to remove FFmpeg web assembly version . I am a novice . Thanks in advance.

**Comment by holzschu** [comment]
`rm ~/Documents/bin/ffmpeg*` should do it.

**Comment by shyam1312** [comment]
thanks . but it returned no file or directory .. need more clarification . i was in correct directory. home folder  . 

need  clarification . does latest version  by default come with two version of ffmpeg .  ( native and wasm ) .    i thought i should delete wasm version to access native . or have to delete wasm  version only if i explicitly installed it in first place .

**Comment by holzschu** [comment]
Latest version comes with only native ffmpeg. If you previously installed WebAssembly version of ffmpeg, it is in your homedirectory and has priority over the native App version. If you never installed it, you have nothing to do.

To check: type `ffmpeg --help`. If it takes several seconds to reply, it is the WebAssembly version. If it is very fast, it is the native version.

If you have the WebAssembly version, it has to be in the $PATH, so: `~/Library/bin` or `~/Documents/bin`.

**Comment by shyam1312** [comment]
ok  thanks . so i tried executing this command . 

$ ffmpeg -i q.mp4 -vcodec libx264 -acodec aac output.mp4

it says unknown encoder 'libx264' . but same thing worked in windows .  ipad screen shot attached . 



 [](url
![IMG_0005](https://user-images.githubusercontent.com/2676950/134151720-dcce4f85-6d52-44dc-9339-1c18e3837d12.PNG)
)

**Comment by holzschu** [comment]
So, if it says `--enable-videotoolbox`, then it is the native version. I'm glad we settled this one.

Next, `libx264` is a GPL library, so it is incompatible with AppStore distribution according to the FSF. You have to use the `h264` codec, which also encodes to H264 format, but using native iOS libraries (also, ffmpeg is pretty good now at guessing the codec from file extensions, so you could just type `ffmpeg -i q.mp4 output.mp4`).

**Comment by shyam1312** [comment]
actually my goal is to convert my videos from 30 fps to 60 fps using  interpolation to create smooth buttery effect . (like Smooth video project  - https://www.svp-team.com/ )  i tried below commands yesterday night. 

ffmpeg -i q.mp4 -vf "minterpolate=fps=60:mi_mode=mci:mc_mode=aobmc:me_mode=bidir:vsbmc=1" output.mp4 . 

ffmpeg -i q.mp4 -filter:v tblend -r 60 result1.mp4

ffmpeg -i q.mp4 -filter:v minterpolate -r 60 result.mp4

but output does acheive 60fps with desired effect . but there is quality loss . although the output file size is higher than input video . but output file is pixelated . 

i am not sure what caused pixelation .  i am gonna try commands again . although this may be out of scope . any help will be gr8 !!

**Comment by Tama47** [comment]
> Latest version comes with only native ffmpeg. If you previously installed WebAssembly version of ffmpeg, it is in your homedirectory and has priority over the native App version. If you never installed it, you have nothing to do.

Where is the native ffmpeg location? I’m trying to use it with yt-dlp --ffmpeg-location option.

**Comment by holzschu** [comment]
I don't understand why you would need to specifify the location of ffmpeg, since it is in the PATH, but it's in `$APPDIR/bin/ffmpeg`.
